sycl-fusion/passes/target/TargetFusionInfo.cpp

```diff
@@ -0,0 +1,267 @@
+//==---------------------- TargetFusionInfo.cpp ----------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TargetFusionInfo.h"
+#include "llvm/ADT/Triple.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/InstrTypes.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsNVPTX.h"
+
+using namespace llvm;
+
+//
+// TargetFusionInfo
+//
+
+TargetFusionInfo TargetFusionInfo::getTargetFusionInfo(llvm::Module *Mod) {
+  llvm::Triple Tri(Mod->getTargetTriple());
+  if (Tri.isNVPTX()) {
+    return TargetFusionInfo(
+        std::shared_ptr<NVPTXTargetFusionInfo>(new NVPTXTargetFusionInfo(Mod)));
+  }
+  if (Tri.isSPIRV() || Tri.isSPIR()) {
+    return TargetFusionInfo(
+        std::shared_ptr<SPIRVTargetFusionInfo>(new SPIRVTargetFusionInfo(Mod)));
+  }
+  assert(false && "Unsupported target for fusion");
+}
+
+//
+// SPIRVTargetFusionInfo
+//
+
+void SPIRVTargetFusionInfo::addKernelFunction(Function *KernelFunc) const {
+  KernelFunc->setCallingConv(CallingConv::SPIR_KERNEL);
+}
+
+ArrayRef<StringRef> SPIRVTargetFusionInfo::getKernelMetadataKeys() const {
+  // NOTE: We do not collect the "kernel_arg_name" metadata, because
+  // the kernel arguments receive new names in the fused kernel.
+  static SmallVector<StringRef> Keys{
+      {"kernel_arg_addr_space", "kernel_arg_access_qual", "kernel_arg_type",
+       "kernel_arg_base_type", "kernel_arg_type_qual"}};
+  return Keys;
+}
+
+void SPIRVTargetFusionInfo::postProcessKernel(Function *KernelFunc) const {
+  // Attach the kernel_arg_name metadata.
+  SmallVector<Metadata *, 16> KernelArgNames;
```

Comment:
```suggestion
  SmallVector<Metadata *> KernelArgNames;
```
Can we rely on the default arg?